src/truncate.jl

```diff
@@ -49,34 +49,81 @@ insupport(d::Truncated{D,Union{Discrete,Continuous}}, x::Real) where {D<:Univari
 
 quantile(d::Truncated, p::Real) = quantile(d.untruncated, d.lcdf + p * d.tp)
 
-for f in [:pdf, :logpdf, :cdf, :logcdf, :ccdf, :logccdf]
-    @eval ($f)(d::Truncated, x::Int) = ($f)(d, float(x))
+function _pdf(d::Truncated, x::T) where {T<:Real}
+    if d.lower <= x <= d.upper
+        pdf(d.untruncated, x) / d.tp
+    else
+        zero(T)
+    end
+end
+
+function pdf(d::Truncated{D}, x::T) where {D<:ContinuousUnivariateDistribution, T<:Real}
+    _pdf(d, x)
+end
+
+function pdf(d::Truncated{D}, x::T) where {D<:DiscreteUnivariateDistribution, T<:Real}
+    isinteger(x) || zero(T)
+    _pdf(d, x)
+end
+
+function pdf(d::Truncated{D}, x::T) where {D<:DiscreteUnivariateDistribution, T<:Integer}
+    _pdf_(d, x)
+end
+
+function _logpdf(d::Truncated, x::T) where {T<:Real}
+    if (d.lower <= x <= d.upper)
+        logpdf(d.untruncated, x) - d.logtp
+    else
+        -Inf
+    end
 end
 
-pdf(d::Truncated, x::T) where {T<:Real} = d.lower <= x <= d.upper ? pdf(d.untruncated, x) / d.tp : zero(T)
+function logpdf(d::Truncated{D}, x::Real) where {D<:DiscreteUnivariateDistribution, T<:Real}
+    isinteger(x) || return -Inf
+    return _logpdf(d, x)
+end
+
+function logpdf(d::Truncated{D}, x::Integer) where {D<:DiscreteUnivariateDistribution}
+    _logpdf(d, x)
+end
 
-logpdf(d::Truncated, x::T) where {T<:Real} = d.lower <= x <= d.upper ? logpdf(d.untruncated, x) - d.logtp : -T(Inf)
+function logpdf(d::Truncated{D, Continuous}, x::T) where {D<:ContinuousUnivariateDistribution, T<:Real}
+    _logpdf(d, x)
+end
 
-cdf(d::Truncated, x::T) where {T<:Real} =
+# fallback to avoid method ambiguities
+_cdf(d::Truncated, x::T) where {T<:Real} =
     x <= d.lower ? zero(T) :
     x >= d.upper ? one(T) :
     (cdf(d.untruncated, x) - d.lcdf) / d.tp
 
-logcdf(d::Truncated, x::T) where {T<:Real} =
-    x <= d.lower ? -T(Inf) :
+cdf(d::Truncated, x::Real) = _cdf(d, x)
+cdf(d::Truncated, x::Integer) = _cdf(d, x) # here for specificity
+
+_logcdf(d::Truncated, x::T) where {T<:Real} =
+    x <= d.lower ? -Inf :
```

Comment:
`T(Inf)` is useful to ensure type stability.

Reply:
ah point, it just broke with integers which don't have `Inf` values, these will be unstable in any case

Reply:
solution might be to always return a `Float64`? The type of probability values does not have to be the same as type of input `x`

Reply:
You can call `float(T)`, that will preserve `Float32`.

Reply:
fixed it on the various functions